Live Migration For Cloud-Hypervisor VMs

kasanchez519 opened the pull request:
Using liuw's branch as a starting point, implement live migration for Cloud-Hypervisor VMs in Libvirt. This entails implementing functionality for Libvirt's 5 main phases of migration: Begin, Prepare, Perform, Finish, and Confirm.

liuw committed:
ch: import migration patch from liw's branch

Signed-off-by: Wei Liu <liuwe@microsoft.com>

Commit:
ch: rebase + compilation fixes

praveen-pk commented on src/ch/ch_conf.c:
```diff
@@ -118,6 +120,8 @@ virCHDriverConfigNew(bool privileged)
 {
     virCHDriverConfig *cfg;
 
+
```
please clean up these white lines

praveen-pk commented on src/libvirt-domain.c:
```diff
@@ -4207,6 +4208,7 @@ virDomainMigrate3(virDomainPtr domain,
     int rc_dst;
     int rc;
 
+    VIR_DEBUG("t-kasanchez DEBUG STATEMENT - virDomainMigrate3 called");
```
please clean up all DEBUG statements with custom prefixes

praveen-pk commented on src/ch/ch_migration.c:
```diff
@@ -230,12 +235,47 @@ chDomainMigrationDstPrepare(virConnectPtr dconn,
     (void) cookieoutlen;
     (void) origname;
 
+    // Start stub Cloud-Hypervisor process on Dest
```
Please clean up these comments

praveen-pk commented on src/ch/ch_driver.c:
```diff
@@ -2060,6 +2064,38 @@ chDomainMigrateFinish3Params(virConnectPtr dconn,
     (void) cancelled;
     return NULL;
+    // virCheckFlags(CH_MIGRATION_FLAGS, NULL);
+    // if (virTypedParamsValidate(params, nparams, CH_MIGRATION_PARAMETERS) < 0)
```
please clean up all the commented out code

praveen-pk commented on src/ch/ch_monitor.c:
```diff
@@ -653,7 +653,7 @@ virCHMonitorPutNoContent(virCHMonitor *mon, const char *endpoint)
     g_autofree char *url = NULL;
     int responseCode = 0;
-    int ret = -1;
+    int ret = 0;
```
Could you please add some detail on why this change is required in the commit message?

praveen-pk commented on src/ch/ch_migration.c:
```diff
@@ -306,7 +307,23 @@ chDomainMigrationSrcPerform(virCHDriver *driver,
+        return -1;
+    
+    cmd = virCommandNew("ch-remote");
+    virCommandAddArgPair(cmd, "--api-socket", "/var/run/libvirt/ch/ch_impish_nfs-socket");
```
Could you follow virCHMonitorNew and derive the socket paths, instead of hardcoding?

praveen-pk commented on src/ch/ch_migration.c:
```diff
@@ -298,6 +298,7 @@ chDomainMigrationSrcPerform(virCHDriver *driver,
                             const char *dname,
                             unsigned int flags)
 {
+    virCommand *cmd;
```
Please use "g_autoptr(virCommand) cmd" here. This automatically frees memory when cmd goes out of scope.

praveen-pk commented on src/ch/ch_migration.c:
```diff
@@ -235,45 +241,78 @@ chDomainMigrationDstPrepare(virConnectPtr dconn,
+    virCommandAddArg(cmd, "-vvv");
+    virCommandAddArgPair(cmd, "--log-file", "/var/log/libvirt/ch.log");
+    VIR_DEBUG("PPK: clh cmd : %s", virCommandToString(cmd, false) );
+    clh = cmd;
```
If this commit only for debugging, please consider dropping it

praveen-pk commented on src/ch/ch_monitor.c:
```diff
@@ -797,16 +797,26 @@ virCHMonitorRefreshThreadInfo(virCHMonitor *mon)
             info[i].type = virCHThreadTypeIO;
             info[i].ioInfo.tid = tids[i];
             virStrcpy(info[i].ioInfo.thrName, data, VIRCH_THREAD_NAME_LEN - 1);
+            VIR_DEBUG("IO Thread, name = %s -> tid: %d",info[i].ioInfo.thrName, tids[i]);
```
Note to praveen, update the commit message with proper context

Commit:
ch: Add Prepare phase for live migration

Commit:
ch: Add Perform phase for live migration

Signed-off-by: Karissa Sanchez <t-kasanchez@microsoft.com>

Commit:
ch: create vm with correct migration features

Update flags to support correct features and have the ch monitor not return errors, since it is currently not fully configured

Signed-off-by: Karissa Sanchez <t-kasanchez@microsoft.com>

Commit:
ch: Implement Finish phase of migration

Signed-off-by: Karissa Sanchez <t-kasanchez@microsoft.com>

Commit:
ch: Implement Confirm phase for live migration

Signed-off-by: Karissa Sanchez <t-kasanchez@microsoft.com>

Commit:
ch: allocate ports for migration

Signed-off-by: Karissa Sanchez <t-kasanchez@microsoft.com>

Commit:
ch: handle ch threads with iou-wrk prefix seperately

cloud-hypervisor started spawning transient thread with iou-wrk threads.
These threads were being tagged as emulator threads. While refreshing threads
disapearance of these threads would result in errors in libvirt.
So, tag these threads as a new class of threads. Doing so, allows the threads
to be transient and libvirt doesn't crash while refreshing emulator threads.

Signed-off-by: Karissa Sanchez <t-kasanchez@microsoft.com>

praveen-pk commented:
Looks Good, Please go ahead and merge.
